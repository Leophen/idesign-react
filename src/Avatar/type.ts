export interface AvatarProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 头像图片地址
   */
  image?: string;
  /**
   * 头像图片 alt 属性
   */
  alt?: string;
  /**
   * 头像大小
   * @default 32
   */
  size?: number;
  /**
   * 头像形状
   * @default circle
   */
  shape?: 'circle' | 'round';
  /**
   * 头像底色
   */
  color?: string;
}

export interface AvatarGroupProps extends AvatarProps {
  /**
   * 折叠方向
   * @default right
   */
  cascading?: "left" | "right";
}

export interface AvatarContextValue {
  inject: (props: AvatarProps) => AvatarProps;
}
