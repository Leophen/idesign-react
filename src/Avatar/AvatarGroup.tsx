import React from 'react';
import classNames from 'classnames';
import './index.scss';
import { AvatarContextValue, AvatarGroupProps, AvatarProps } from './type'

export const AvatarContext = React.createContext<AvatarContextValue>(null as any);

const AvatarGroup: React.FC<AvatarGroupProps> = (props) => {
  const {
    children,
    className,
    style,
    size = 32,
    shape = "circle",
    color,
    cascading = "right",
  } = props

  // 注入每一项的 context
  const context: AvatarContextValue = {
    // 将头像组的 props 注入单个头像的方法
    inject: (singleAvatarProps: AvatarProps) => {
      return {
        size,
        shape,
        color,
        ...singleAvatarProps,
      };
    },
  };

  return (
    <AvatarContext.Provider value={context}>
      <div
        className={classNames(
          'i-avatar-group',
          `i-avatar__offset-${cascading}`,
          className
        )}
        style={{ ...style }}
      >
        {children}
      </div>
    </AvatarContext.Provider>
  )
}

export default AvatarGroup;
