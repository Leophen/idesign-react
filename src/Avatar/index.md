---
nav:
  title: 组件
  path: /components
group:
  title: 数据展示组件
  order: 4
order: 4
---

# Avatar 头像

用于展示头像的组件。

## 基本用法

通过 `image` 属性传入头像地址，如果链接失效或不传入链接会显示默认用户图标：

```tsx
import React from 'react';
import { Avatar } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Avatar image="https://picsum.photos/180/120" />
      <Avatar image="https://picsum123.photos/180/120" />
      <Avatar />
    </div>
  );
};

export default App;
```

## 自定义头像内容

通过 `children` 子节点自定义头像内容：

```tsx
import React from 'react';
import { Avatar } from 'idesign-react';

const App = () => {
  return (
    <Avatar>L</Avatar>
  );
};

export default App;
```

## 自定义头像大小

通过 `size` 属性控制头像大小：

```tsx
import React from 'react';
import { Avatar } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Avatar size={24}>L</Avatar>
      <Avatar size={32}>L</Avatar>
      <Avatar size={40}>L</Avatar>
    </div>
  );
};

export default App;
```

## 自定义头像形状

通过 `shape` 属性控制头像形状：

```tsx
import React from 'react';
import { Avatar } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Avatar shape="circle">L</Avatar>
      <Avatar shape="round">L</Avatar>
    </div>
  );
};

export default App;
```

## 自定义头像底色

通过 `color` 属性控制头像底色：

```tsx
import React from 'react';
import { Avatar } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Avatar color="#3236A4">L</Avatar>
      <Avatar color="#8AD2B8">L</Avatar>
    </div>
  );
};

export default App;
```

## 组合头像

可通过 `Avatar.Group` 包裹为组合头像：

```tsx
import React from 'react';
import { Avatar } from 'idesign-react';

const App = () => {
  return (
    <Avatar.Group>
      <Avatar>L</Avatar>
      <Avatar>P</Avatar>
      <Avatar>L</Avatar>
    </Avatar.Group>
  );
};

export default App;
```

## 组合折叠方向

可通过 `cascading` 属性控制组合头像折叠方向：

```tsx
import React from 'react';
import { Avatar } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <Avatar.Group cascading="right">
        <Avatar>L</Avatar>
        <Avatar>P</Avatar>
        <Avatar>L</Avatar>
      </Avatar.Group>
      <Avatar.Group cascading="left">
        <Avatar>L</Avatar>
        <Avatar>P</Avatar>
        <Avatar>L</Avatar>
      </Avatar.Group>
    </div>
  );
};

export default App;
```

## 组合头像透传

`Avatar.Group` 可透传 `size`、`shape`、`color` 属性，优先级低于单个头像对应的属性：

```tsx
import React from 'react';
import { Avatar } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <Avatar.Group
        size={40}
        shape="round"
        color="#5598F7"
      >
        <Avatar>L</Avatar>
        <Avatar>P</Avatar>
        <Avatar>L</Avatar>
      </Avatar.Group>
      <Avatar.Group
        size={40}
        shape="round"
        color="#5598F7"
      >
        <Avatar size={32}>L</Avatar>
        <Avatar shape="circle">P</Avatar>
        <Avatar color="#F0B9B7">L</Avatar>
      </Avatar.Group>
    </div>
  );
};

export default App;
```

## Avatar API

<API hideTitle />

## AvatarGroup API

| 属性      | 说明         | 类型            | 默认值    |
| --------- | ------------ | --------------- | --------- |
| children  | 头像         | `ReactNode`     | `--`        |
| className | 组合头像类名 | `string`        | `--`        |
| style     | 组合头像样式 | `CSSProperties` | `--`        |
| size      | 全局头像大小 | `number`        | `32`      |
| shape     | 全局头像形状 | `circle〡round` | `circle`  |
| color     | 全局头像底色 | `string`        | `#c1c4cb` |
| cascading | 组合折叠方向 | `left〡right`   | `right`   |
