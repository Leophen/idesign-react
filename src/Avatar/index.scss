.i-avatar {
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  overflow: hidden;
  color: #fff;
  background: var(--i-font-5);
  border-radius: 50%;
  &.i-avatar__shape-round {
    border-radius: 6px;
  }
}

.i-avatar__image {
  width: 100%;
  height: 100%;
}

.i-avatar-group {
  display: flex;
  align-items: center;
  box-sizing: border-box;
  .i-avatar {
    border: 2px solid var(--i-bg);
  }
  &.i-avatar__offset-right {
    .i-avatar:not(:last-child) {
      margin-right: -6px;
    }
  }
  &.i-avatar__offset-left {
    .i-avatar {
      &:first-child {
        z-index: 49;
      }
      &:nth-child(2) {
        z-index: 48;
      }
      &:nth-child(3) {
        z-index: 47;
      }
      &:nth-child(4) {
        z-index: 46;
      }
      &:nth-child(5) {
        z-index: 45;
      }
      &:nth-child(6) {
        z-index: 44;
      }
      &:nth-child(7) {
        z-index: 43;
      }
      &:nth-child(8) {
        z-index: 42;
      }
      &:nth-child(9) {
        z-index: 41;
      }
      &:nth-child(10) {
        z-index: 40;
      }
      &:nth-child(11) {
        z-index: 39;
      }
      &:nth-child(12) {
        z-index: 38;
      }
      &:nth-child(13) {
        z-index: 37;
      }
      &:nth-child(14) {
        z-index: 36;
      }
      &:nth-child(15) {
        z-index: 35;
      }
      &:nth-child(16) {
        z-index: 34;
      }
      &:nth-child(17) {
        z-index: 33;
      }
      &:nth-child(18) {
        z-index: 32;
      }
      &:nth-child(19) {
        z-index: 31;
      }
      &:nth-child(20) {
        z-index: 30;
      }
      &:nth-child(21) {
        z-index: 29;
      }
      &:nth-child(22) {
        z-index: 28;
      }
      &:nth-child(23) {
        z-index: 27;
      }
      &:nth-child(24) {
        z-index: 26;
      }
      &:nth-child(25) {
        z-index: 25;
      }
      &:nth-child(26) {
        z-index: 24;
      }
      &:nth-child(27) {
        z-index: 23;
      }
      &:nth-child(28) {
        z-index: 22;
      }
      &:nth-child(29) {
        z-index: 21;
      }
      &:nth-child(30) {
        z-index: 20;
      }
      &:nth-child(31) {
        z-index: 19;
      }
      &:nth-child(32) {
        z-index: 18;
      }
      &:nth-child(33) {
        z-index: 17;
      }
      &:nth-child(34) {
        z-index: 16;
      }
      &:nth-child(35) {
        z-index: 15;
      }
      &:nth-child(36) {
        z-index: 14;
      }
      &:nth-child(37) {
        z-index: 13;
      }
      &:nth-child(38) {
        z-index: 12;
      }
      &:nth-child(39) {
        z-index: 11;
      }
      &:nth-child(40) {
        z-index: 10;
      }
      &:nth-child(41) {
        z-index: 9;
      }
      &:nth-child(42) {
        z-index: 8;
      }
      &:nth-child(43) {
        z-index: 7;
      }
      &:nth-child(44) {
        z-index: 6;
      }
      &:nth-child(45) {
        z-index: 5;
      }
      &:nth-child(46) {
        z-index: 4;
      }
      &:nth-child(47) {
        z-index: 3;
      }
      &:nth-child(48) {
        z-index: 2;
      }
      &:nth-child(49) {
        z-index: 1;
      }
      &:nth-child(50) {
        z-index: 0;
      }
      &:not(:first-child) {
        margin-left: -6px;
      }
    }
  }
}
