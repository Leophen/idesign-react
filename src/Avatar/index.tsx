import React, { useContext, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import AvatarGroup, { AvatarContext } from './AvatarGroup'
import { AvatarProps } from './type';

const Avatar: React.FC<AvatarProps> & { Group: React.ElementType } = (props) => {
  // 存在头像组时从 Context 注入全局属性
  const context = useContext(AvatarContext);
  const newProps = context ? context.inject(props) : props;

  const {
    children,
    className,
    style,
    image,
    size = 32,
    shape = "circle",
    color,
    ...restProps
  } = newProps;

  const [imgLoad, setImgLoad] = useState(true)
  const handleImgLoadError = () => {
    setImgLoad(false)
  }

  return (
    <div
      className={classNames(
        'i-avatar',
        shape === 'round' && `i-avatar__shape-${shape}`,
        className
      )}
      style={{ ...style, width: size, height: size, background: color }}
      {...restProps}
    >
      {image && imgLoad ?
        <img
          className='i-avatar__image'
          src={image}
          onError={handleImgLoadError}
          referrerPolicy="no-referrer"
        /> :
        (children ? children : <Icon name="User" color="#fff" />)
      }
    </div>
  );
};

Avatar.Group = AvatarGroup

AvatarGroup.displayName = 'AvatarGroup';
Avatar.displayName = 'Avatar';

export default Avatar;
