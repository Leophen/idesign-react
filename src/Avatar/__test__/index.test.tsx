import { render } from '@testing-library/react';
import Avatar from '../index';

const imageSrc = 'https://picsum.photos/180/120';

describe('Avatar 组件测试', () => {
  const renderAvatar = (avatar) => {
    const { container } = render(avatar);
    return container.firstChild;
  }

  it('image alt', () => {
    const { getByAltText } = render(<Avatar image={imageSrc} alt="test-avatar" />);
    const image = getByAltText('test-avatar');
    expect(image.src).toBe(imageSrc);
    const { asFragment } = render(<Avatar image={imageSrc} alt="test-avatar" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('children', () => {
    const { queryByText } = render(<Avatar>L</Avatar>);
    expect(queryByText('L')).toBeInTheDocument();
  });

  it('size', () => {
    expect(renderAvatar(<Avatar size={24}>L</Avatar>)).toHaveStyle('width: 24px');
  })

  it('shape', () => {
    expect(renderAvatar(<Avatar shape="round">L</Avatar>)).toHaveClass('i-avatar__shape-round');
  })

  it('color', () => {
    expect(renderAvatar(<Avatar color="#3236A4">L</Avatar>)).toHaveStyle('background: #3236A4');
  })
});

