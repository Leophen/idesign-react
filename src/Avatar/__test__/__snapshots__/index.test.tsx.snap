// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar 组件测试 image alt 1`] = `
<DocumentFragment>
  <div
    class="i-avatar"
    style="width: 32px; height: 32px;"
  >
    <img
      alt="test-avatar"
      class="i-avatar__image"
      referrerpolicy="no-referrer"
      src="https://picsum.photos/180/120"
    />
  </div>
</DocumentFragment>
`;
